graphql_compiler/cost_estimation/analysis.py

```diff
@@ -0,0 +1,255 @@
+        distinct_result_set_estimates: see get_distinct_result_set_estimates
+        query_metadata: info on locations, inputs, outputs, and tags in the query
+        parameters: the query parameters
+
+    Returns:
+        the pagination capacity of each PropertyPath
+    """
+    pagination_capacities = {}
+    for location, location_info in query_metadata.registered_locations:
+        vertex_type_name = location_info.type.name
+
+        for field_name, _ in location_info.type.fields.items():
+            property_path = PropertyPath(location.query_path, field_name)
+            if not is_meta_field(field_name):
+                if is_uuid4_type(schema_info, vertex_type_name, field_name):
+                    pagination_capacities[property_path] = distinct_result_set_estimates[
+                        location.query_path
+                    ]
+                elif field_supports_range_reasoning(schema_info, vertex_type_name, field_name):
+                    field_value_interval = field_value_intervals.get(
+                        property_path, Interval(None, None)
+                    )
+                    quantiles = schema_info.statistics.get_field_quantiles(
+                        vertex_type_name, field_name
+                    )
+                    if quantiles is not None:
+
+                        # The first and last values of the quantiles are the minimum and maximum
+                        # observed values. We call all other values the proper quantiles. We don't
+                        # directly use the minimum and maximum values as page boundaries since we
+                        # will most likely generate empty pages.
+                        proper_quantiles = quantiles[1:-1]
+
+                        # Get the relevant quantiles (ones inside the field_value_interval)
+                        min_quantile = 0
+                        max_quantile = len(proper_quantiles)
+                        if field_value_interval.lower_bound is not None:
+                            min_quantile = bisect.bisect_left(
+                                proper_quantiles, field_value_interval.lower_bound
+                            )
+                        if field_value_interval.upper_bound is not None:
+                            max_quantile = bisect.bisect_left(
+                                proper_quantiles, field_value_interval.upper_bound
+                            )
+                        relevant_quantiles = proper_quantiles[min_quantile:max_quantile]
+
+                        # TODO(bojanserafimov): If the relevant quantiles contain duplicates, the
+                        #                       pagination capacity would be lower.
+
+                        pagination_capacities[property_path] = min(
+                            len(relevant_quantiles) + 1,
+                            distinct_result_set_estimates[location.query_path],
+                        )
```

Comment:
This should probably be extracted into one or more functions. For example, you could make the `elif` block be a function call, and then outside that `if/elif` (i.e. at the same indentation level as line 164) you could write something like: 
```python
if pagination_capacity_for_path is not None:
    pagination_capacities[property_path] = pagination_capacity_for_path
```

I think that would make this much cleaner and easier to read.